JSON.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+#include <string>
+#include <map>
+#include <istream>
+#include <exception>
+
+#include <iostream>
+
+
+/**
+ * \class JSON
+ *
+ * \brief JSON class
+ *
+ */
+class JSON
+{
+	std::map<std::string, std::string> scenario;
+public:
+
+	JSON()
+	{
+
+	}
+
+
+	JSON(std::map<std::string, std::string> scen) : scenario(scen)
+	{
+
+	}
+
+	/**
+	* \brief Parsing from an istream
+	*
+	* \param istream with JSON-like content
+	*
+	* \return Map of key-value pairs
+	*/
+	static std::map<std::string, std::string> parseFromIstream(std::istream& instream); //istream input
+
+
+	/**
+	* \brief Parsing from a file
+	*
+	* \param string JSON textfile's name
+	*
+	* \return Map of key-value pairs
+	*/
+	static std::map<std::string, std::string> parseFromFile(const std::string& fname); //filename input
+
+	/**
+	* \brief Parsing from a string
+	*
+	* \param string with JSON-like content
+	*
+	* \return Map of key-value pairs
+	*/
+	static std::map<std::string, std::string> parseFromString(std::string& data); //string input
+
+
+	JSON operator= (std::map<std::string, std::string>& other)
+	{
+		JSON json(other);
+		return json;
+	}
+
+
+	template<typename T>
+	T get(const std::string& getFile)
+	{
+		if(getFile == "hero")
+		{
+			return scenario["hero"];
+		}
+		else if (getFile == "monsters")
+		{
+			return scenario["monsters"];
+		}
+
+		return "";
+
+	}
```

Comment:
Mukodik a megadott main-nel, szoval "technikailag" bemehetne (es kesobb szivnatok rendesen vele), de van idotok, ugyhogy ezt tegyetek rendbe. Itt a template lenyeget nem sikerult megerteni. A template lenyege az, hogy ez mukodjon T=int-re is. Nektek ez tovabbra is csak string-re fog mukodni, kulonben forditasi hibat ad az egesz. 

Annak semmi ertelme, hogy a getfile-ra csinalsz egy if-et,  siman at(getFile) lehetett volna.

Az meg vegtelenul csunya, hogy van egy JSON parser, amiben monster meg hero szerepelnek. Mi koze egy json-nek ilyenekhez?!  meg scenario. A JSON az JSON-r parzol. Ezt a kodot modositas nelkul at kellene tudni tenni egy webalkalmazasba, ahol felhasznaloi adatokat parzolna be, vagy akarmi.

Reply:
@hegyhati 
A problémát értem, a megoldás kitalálása viszont már nem megy. Odáig világos, hogy ha már van ez a template `get `függvényem, akkor sokkal egyszerűbb lekérni a JSON-től bármit. Viszont egész idáig a parser mindent string-ként tárolt és a felhasználási helyén konvertáltuk át arra, amire éppen szükséges. Erre épült az egész.
Mivel nem castolhatom át, hogy `(T) content.at(key)` (ahol content a map), nem tudom hogy hogyan tudnám megcsinálni anélkül, hogy ismét átdolgoznám a parsert, ráadásul a map is legalább háromféle típust kéne hogy tároljon az értékeiben (std::variant?). Vagy a template paraméter típusára kéne rávizsgálnom?
Lehet, hogy valami teljesen triviális dolgot nem veszek észre, de most teljesen megakadtam.

Reply:
Igen, nem is baj, hogy elsore `map<string,string>` volt az adatszerkezet, es utana castoltatok. De nyilvan Ez nem a legjobb megoldas, foleg ha majd ugye megyunk tovabb arra, hogy mondjuk lesz array is. (a main-ben amugy nagyon jol jott volna, de "megkerultem"). Igaz akkor mar megint egy picit mashogy allnank hozza.

Visszakanyarodva: az `std::variant` egy teljesen jo megoldas (en magamnak azzal csinaltam meg), de sima OO-n az nem volt. Ha voltal Adam targyan, vagy megtanulod hasznalni Magadtol, akkor tokeletes.  A nem type safe megoldasa a dolognak ugye a `union`. Meg a regi C-s idokbol. 

Amit mindenkepp takarits ki innet, az a use-cas especifdic (warrior/scenario) stb dolgok. Aztan en azt javasolnam, hogy agaztass le errol egy branchet, amin atirod std::variant-ra. Ha szuperul mukodik, merge-eld bele ebbe, es minden oke. Ha nem, bemegy majd ez a masterbe. De a JSON-bol tenyleg keruljon ki minden, ami nem JSON.

Reply:
@hegyhati 
A JSON-ből már el is tüntettem a szükségtelen dolgokat / átneveztem a scenario-t content-nek. Azt akkor fel is push-olom ide.

Meg akartuk kerülni a dolgokat azzal, hogy először csináljuk a refactort és csak utána állunk neki a többi kisebb feladatnak, hogy azokat már egyáltalán ne kelljen módosítani később. Nem vagyok biztos benne, hogy még szerda előtt be tudnám fejezni az átírást, hogy még a többi feladatra is legyen időnk.
Csinálhatunk olyat, hogy ebből a refactor branch-ből ágazunk le a többivel, és ezzel (vagy a mostani, vagy a variantos állapotával) együtt mennek be a masterbe, vagy szívjunk tisztességes ember módjára és kezdjük masterből az egészet a régi file nevekkel, változókkal stb.?

Reply:
@hegyhati
Az előző kommentem tárgytalan. Viszonylag fájdalommentes volt az áttérés variant-ra.
Elvileg elkészültem a kért módosításokkal.

Reply:
No, orulok, hogy megoldodott :-)  pont azert vannak ezek a dolgok, hogy fajdalommentesen tudjunk ezt-azt csinalni :-) nezem.